addons/website_sale/controllers/main.py

```diff
@@ -1830,6 +1830,52 @@ def products_recently_viewed_delete(self, product_id=None, product_template_id=N
             request.env['website.track'].sudo().search(domain).unlink()
         return {}
 
+    @route('/shop/categories', type='jsonrpc', auth='public', website=True)
+    def get_shop_categories(self, category_id=None):
+        """
+        Return a list of public product categories available on the website.
+
+        If `category_id` is provided, returns the children of that category.
+        If no children are found, the category itself is returned.
+        If `category_id` is not provided, top-level categories are returned.
+
+        :param category_id: (optional) ID of the parent category to fetch children for
+        :returns: A list of dictionaries representing the filtered product categories.
+              Each dictionary contains:
+                  - `id` (int): The category's ID
+                  - `name` (str): The category's display name
+                  - `cover_image_url` (str): image URL for the category
+        :rtype: list[dict]
+        """
+        if not request.website.has_ecommerce_access():
+            return request.redirect('/web/login')
+
+        CategorySudo = request.env['product.public.category'].sudo()
+        domain = CategorySudo._get_available_category_domain(request.website.id)
+        if not category_id:
+            categories = CategorySudo.search(domain + [('parent_id', '=', False)])
+        else:
+            parent = CategorySudo.search(domain + [('id', '=', category_id)])
+            categories = (
+                parent.child_id.filtered_domain(domain) or parent
+            )
+        return [{
+            'id': cat.id,
+            'name': cat.name,
+            'cover_image_url': (
+                f'{request.website.get_base_url()}{request.website.image_url(cat, "cover_image")}'
+                if cat.cover_image
+                else request.env['product.template']._get_product_placeholder_filename()
+            ),
+        } for cat in categories]
+
+    @route('/snippets/category/set_image', type='jsonrpc', auth='user')
+    def set_category_image(self, category_id, media):
+        image_data = request.env['ir.attachment'].browse(media[0]['id']).datas
+        category = request.env['product.public.category'].browse(category_id)
```

Comment:
browse can return the record with id even if it does not exist
```suggestion
        category = request.env['product.public.category'].browse(category_id).exists()
```